persist: commit state updates to durable storage incrementally

danhhz opened the pull request:
Before, there was pressure to keep the size of state down, because it
was rewritten entirely on each command application. In particular, this
created a tension in compaction tuning between being aggressive about
fewer batches (smaller state) and compacting small batches lazily
(smaller write amplification).

Writing state updates as incremental diffs means that size of a
Consensus writes for each command is independent of the total size of
state. We should be able leverage this to make the entire
`WriteHandle::compare_and_append_batch` latency constant w.r.t. the size
of state and thus independent of compaction. This lets us tune
compaction entirely for where we want to be in its more intrinsic
tradeoff between read, write, and space amplification.

(NB: This commit doesn't quite get us to constant latencies, there's
some elbow grease left. I've proven concretely that it can get down to
`O(log(num state batches))`, but that included some hacks that didn't
make this PR. This would be lovely followup work once we get a chance.)

As persist metadata changes over time, we make its versions (each
identified by a [SeqNo]) durable in two ways:
- `rollups`: Periodic copies of the entirety of [State], written to
  [Blob].
- `diffs`: Incremental [StateDiff]s, written to [Consensus]. The
following invariants are maintained at all times:
- A shard is initialized iff there is at least one version of it in
  Consensus.
- The first version of state is written to `SeqNo(1)`. Each successive
  state version is assigned its predecessor's SeqNo +1.
- `current`: The latest version of state. By definition, the largest
  SeqNo present in Consensus.
- As state changes over time, we keep a range of consecutive versions
  available. These are periodically `truncated` to prune old versions
  that are no longer necessary.
- `earliest`: The first version of state that it is possible to
  reconstruct.
  - Invariant: `earliest <= current.seqno_since()` (we don't garbage
    collect versions still being used by some reader).
  - Invariant: `earliest` is always the smallest Seqno present in
    Consensus.
    - This doesn't have to be true, but we select to enforce it.
    - Because the data stored at that smallest Seqno is an incremental
      diff, to make this invariant work, there needs to be a rollup at
      either `earliest-1` or `earliest`. We choose `earliest` because it
      seems to make the code easier to reason about in practice.
    - A consequence of the above is when we garbage collect old versions
      of state, we're only free to truncate ones that are `<` the latest
      rollup that is `<= current.seqno_since`.
- `live diffs`: The set of SeqNos present in Consensus at any given
  time.
- `live states`: The range of state versions that it is possible to
  reconstruct: `[earliest,current]`.
  - Because of earliest and current invariants above, the range of `live
    diffs` and `live states` are the same.
- The set of known rollups are tracked in the shard state itself.
  - For efficiency of common operations, the most recent rollup's Blob
    key is always denormalized in each StateDiff written to Consensus.
    (As described above, there is always a rollup at earliest, so we're
    guaranteed that there is always at least one live rollup.)
  - Invariant: The rollups in `current` exist in Blob.
    - A consequence is that, if a rollup in a state you believe is
      `current` doesn't exist, it's a guarantee that `current` has
      changed (or it's a bug).
  - Any rollup at a version `< earliest-1` is useless (we've lost the
    incremental diffs between it and the live states). GC is tasked with
    deleting these rollups from Blob before truncating diffs from
    Consensus. Thus, any rollup at a seqno < earliest can be considered
    "leaked" and deleted by the leaked blob detector.
  - Note that this means, while `current`'s rollups exist, it will be
    common for other live states to reference rollups that no longer
    exist.

### Motivation

  * This PR adds a known-desirable feature.

### Tips for reviewer

<!--
Leave some tips for your reviewer, like:

    * The diff is much smaller if viewed with whitespace hidden.
    * [Some function/module/file] deserves extra attention.
    * [Some function/module/file] is pure code movement and only needs a skim.

Delete this section if no tips.
-->

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - N/A

pH14 commented on src/persist-client/src/internal/state_fetcher.rs:
```diff
@@ -0,0 +1,373 @@
+    }
+
+    // WIP Returns all diffs and a rollup at the version == the first diff
+    pub async fn fetch_live_states(
```
since Fetcher is in the struct name, maybe just states? e.g. `self.fetcher.states(shard_id)` I'm also not sure we need live anywhere, since I don't think the fetcher can return non-live values?

danhhz replied:
IMO the live is pretty important. I've defined "live" since you wrote this, so maybe it makes more sense now? it's no longer called Fetcher, so I think we need the verb now

pH14 commented on src/persist-client/src/internal/state_fetcher.rs:
```diff
@@ -0,0 +1,373 @@
+    pub async fn fetch_live_states(
+        &self,
+        shard_id: &ShardId,
+    ) -> Result<(State<K, V, T, D>, Vec<VersionedData>), CodecMismatch> {
```
if there's a tolerable way to handle the lifetimes, I think the API / usage would be much cleaner if we returned something like `StateIterator<K, V, T, D>` that internally owns the state + diffs, and produces &States as an iterator

danhhz replied:
done!

there are tricks to impl Iterator over references, but they rely on multiple non-mut refs (long story). since we have to mutate state here, they can't work. luckly, we don't actually have to impl Iterator, `while let Some` is just as good for the two places we use this

pH14 commented on src/persist-client/src/internal/state_fetcher.rs:
```diff
@@ -0,0 +1,373 @@
+                    // Initialize it and try again. This throws away the return
+                    // value of maybe_init, but that's fine because we should
+                    // never hit this case in practice.
+                    let _ = self.maybe_init(shard_id).await?;
```
could potentially avoid the are-we-init-or-not questions by creating two Fetcher types. calling `maybe_init` on the first returns a Fetcher that can assume its shard is always initialized. IMO this isn't total bikeshedding, I think it could clarify the logic in here a bit

danhhz replied:
I think this one is just a confusing comment, i've replaced it with a panic, which is what I should have done in the first place. the only method in this whole file that even considers an initialized shard, is maybe_init_shard (and now try_compare_and_set_current)

I _think_ we could also make Fetcher::new into something that returns `(Self, State<K, V, T, D>)` and then Machine::new would take State as an argument (guaranteeing that Machine is only ever dealing with initialized shard) but I haven't gotten there yet

danhhz replied:
> I think we could also make Fetcher::new into something that returns (Self, State<K, V, T, D>) and then Machine::new would take State as an argument (guaranteeing that Machine is only ever dealing with initialized shard) but I haven't gotten there yet

played with this for a bit, but couldn't manage anything that didn't feel like a significant readability regression to me. I added an explicit mention in the rustdoc for each method whether it can be called on an uninitialized shard or not, maybe that helps?

pH14 commented on src/persist-client/src/internal/state_fetcher.rs:
```diff
@@ -0,0 +1,373 @@
+    }
+
+    // WIP only callable on initialized shards, so panics if hint is empty
+    async fn fetch_latest_state(
```
so tbh I've spent quite a lot of time trying to understand StateFetcher and have gotten very turned about a few times, and I think some of it is just naming. I find this one tricky because it doesn't fetch the _latest_ state, it constructs the latest state up through `hint`, and what hint is isn't obvious from the type.

this might be easier to parse if it were something like:

```
async fn state(&self, shard_id, diffs: Vec<StateDiff>)
```

danhhz replied:
I've tried to solve this one with rustdoc, too. grok-ability matters a great deal to me, but to be entirely honest, I've probably passed 32 hours of tweaking state_fetcher/state_versions at this point, and run out of motivation for more tweaks. i'd love to discuss a concrete proposal (i.e. a diff that passes CI) if you have one <3

pH14 commented on src/persist-client/src/internal/state_fetcher.rs:
```diff
@@ -0,0 +1,373 @@
+///   - it follows that there must be a rollup at the smallest live seqno or the
+///     smallest live seqno-1. we enforce the former
+/// - rollups for seqnos < smallest live seqno are leaked
+///   - this means not all rollups referenced by live seqnos exist
```
hmm not sure I follow how the statement implies the second (also leaked isn't defined, so not quiiiiite sure I have the same meaning in my head)

danhhz replied:
this is all precisely defined now

pH14 commented on src/persist-client/src/internal/state_fetcher.rs:
```diff
@@ -0,0 +1,373 @@
+///
+/// WIP Invariants:
+/// - diffs for current.seqno_since..=current exist
+/// - all live seqnos are constructable
```
might be helpful to have a quick definition on live

danhhz replied:
done!

pH14 commented on src/persist-client/src/internal/gc.rs:
```diff
@@ -181,58 +216,113 @@ impl GarbageCollector {
+        // anything before it.
+        //
+        // It's unfortunate to tie the rollups schedule to gc so tightly, but
+        // the above invariant makes our leaked blob logic easier.
```
if GC is held back for a long time, would we ever want to force a state rollup to be written? otherwise new handles may need to run arbitrarily long scans through consensus to reconstruct the latest state if the seqno since is held back

pH14 replied:
also I think I'm missing a key insight into how this invariant makes leaked blob detection simpler 😓

danhhz replied:
> if GC is held back for a long time, would we ever want to force a state rollup to be written? otherwise new handles may need to run arbitrarily long scans through consensus to reconstruct the latest state if the seqno since is held back

yeah, but I think we have to cut scope at some point and get this merged. adding a separate schedule for rollups is inevitable, but it's a decent chunk of work and I'm not sure it's worth missing this next release over

> also I think I'm missing a key insight into how this invariant makes leaked blob detection simpler 😓

explained in the big StateVersions rustdoc (and a pointer to it put here). summary: any rollup < the earliest seqno is consensus has been leaked

danhhz replied:
> I think we can just put this as another Maintenance task that triggers when there's a big delta between latest_rollup and seqno
super straightforward. done!

pH14 reviewed (commented):
Yay! This is great. I didn't mean to do a line-level review, but that's a bit what happened when I tried to really grok what StateFetcher is solving.

The big picture here makes a lot of sense to me, and I think the main hangups I had reading through are just around naming and some of the code structure, with one open question on the rollup creation schedule. Seems like if we get the abstraction of the StateFetcher just so, the impact on the existing code should be quite minimal!

danhhz commented on src/persist-client/src/internal/machine.rs:
```diff
@@ -578,16 +614,19 @@ where
                         cas_mismatch_metric.0.inc();
-                        self.update_state(current).await;
+                        self.fetcher
+                            .update_state(&mut self.state, live_diffs)
```
done! (mostly)

moved the apply methods to StateDiff instead of creating a new StateApplier. deconstructed update_state into roughly the pattern you mentioned (replacing Option<()> will bool 😂). still need some polish on fetch_latest_state, it's harder than you'd initially guess for reasons I'm too tired to explain today

danhhz replied:
explained the difference between fetch_current_state and the new fetch_and_update_to_current in rustdoc (one is all live diffs, the other is only diffs between some in-mem state.seqno and current)

danhhz commented:
responded to the comments I've addressed, the other ones are still WIP

danhhz commented on src/persist-client/src/internal/state_diff.rs:
```diff
@@ -0,0 +1,528 @@
+fn sniff_compaction<'a, T: Timestamp + Lattice>(
+    diffs: &'a [StateFieldDiff<HollowBatch<T>, ()>],
+) -> Option<(Vec<&'a HollowBatch<T>>, HollowBatch<T>)> {
+    // Grab the first and only Insert.
```
yup, reworded

danhhz commented on src/persist-client/src/internal/state_diff.rs:
```diff
@@ -0,0 +1,528 @@
+        StateFieldValDiff::Insert(()) => Some(&x.key),
+        _ => None,
+    });
+    let insert = match inserts.next() {
```
renamed "insert" to "compaction_output" and "deletes" to "compaction_inputs"

danhhz commented:
okay, I've addressed all comments (I think), added docs and comment and got a green CI build. i'll try to write some basic tests tomorrow, but tbh my motivation for the week is running out and it's just as easy to write tests as a followup PR on monday, so I'm not promising anything. this should be ready for final review!

danhhz commented on src/persist-client/src/internal/state.proto:
```diff
@@ -60,9 +58,40 @@ message ProtoStateRollup {
+    uint64 seqno_to = 3;
+    string latest_rollup_key = 4;
+
+    // WIP columnar encoding of these just in case?
```
still haven't gotten around to prototyping this. worst case, we can change it right before oct 11

pH14 reviewed (approved):
ooo the `StateApplier` and its usage reads much more easily than before, and the added rustdoc and definitions are 👌. 

I spent most of my time in `Machine`, `StateApplier`, and `GC` and the logic in those LGTM. I'm less confident in my abilities to grok all the implications of the different paths in `StateDiff` but on the surface they seem reasonable. Let's do it!

danhhz commented on src/persist-client/src/internal/paths.rs:
```diff
@@ -51,48 +53,115 @@ impl PartId {
 
-impl std::fmt::Display for PartialBlobKey {
+/// Partially encoded path used in [mz_persist::location::Blob] storage.
+/// Composed of a [crate::write::WriterId] and [crate::internal::paths::PartId].
```
done!

danhhz commented on src/persist-client/src/internal/state_diff.rs:
```diff
@@ -0,0 +1,576 @@
+}
+
+#[allow(dead_code)]
+fn sniff_compaction<'a, T: Timestamp + Lattice>(
```
for me, that feels a bit too much in the "log of commands" direction, but this this method has come up with question from you in both of your review passes, which I think is a signal worth paying attention to

I don't plan on re-enabling sniff_compaction before merge (falling back to the slow path for compaction feels quite a bit less risky given that's the path that's always used pre-inc-state, and this PR wants any risk reduction it can get), so left resolving how to do this as a TODO

aljoscha reviewed (commented):
First bunch of comments. I got until about half of the final (big) commit.

danhhz commented on src/persist/src/location.rs:
```diff
@@ -646,20 +648,22 @@ pub mod tests {
         };
 
-        // Trying to update without the correct expected seqno fails, (even if expected > current)
+        // Trying to update without the correct expected seqno succeeds, (even
```
uhhh no idea what happened here. fixed

danhhz commented on src/persist-client/src/internal/encoding.rs:
```diff
@@ -341,9 +337,8 @@ impl<T: Timestamp + Lattice + Codec64> RustType<ProtoTrace> for Trace<T> {
             // tune things (especially when we add inc state) the rate of
             // this deserialization should go down. Revisit as necessary.
-            ret.push_batch(batch);
+            let _ = ret.push_batch(batch);
```
done!

danhhz commented on src/persist-client/src/internal/machine.rs:
```diff
@@ -574,21 +602,45 @@ where
+                            )
+                        } else {
+                            // Otherwise, we've gc'd the diffs we'd need to advance state to where
+                            // hint is at. Use the hint to kick off fetching a new rollup and use
```
oh wow, I went to update this comment and in typing it out, realized that (95% sure) in this branch diffs_to_current actually _is_ all live diffs and we don't need to fetch it again. great find!

danhhz commented on src/persist-client/src/internal/machine.rs:
```diff
@@ -600,99 +652,18 @@ where
-        debug_assert_eq!(current.seqno, current_state.seqno());
-        debug_assert!(self.state.seqno() <= current.seqno);
-        self.state = current_state;
+        let seqno_before = self.state.seqno;
```
yeah this gets called a lot :flushed:

aljoscha commented on src/persist-client/src/internal/state.proto:
```diff
@@ -60,9 +58,40 @@ message ProtoStateRollup {
     string diff_codec = 5;
     uint64 seqno = 6;
     uint64 last_gc_req = 10;
+    map<uint64, string> rollups = 12;
```
Funny that `ProtoStateRollup` now has a `rollups` field. Maybe the name is not accurate anymore. 😅

aljoscha commented on src/persist-client/src/internal/state_diff.rs:
```diff
@@ -0,0 +1,576 @@
+            latest_rollup_key.clone(),
+        );
+        diff_field_single(from_last_gc_req, to_last_gc_req, &mut diffs.last_gc_req);
+        diff_field_sorted_iter(from_rollups.iter(), to_rollups, &mut diffs.rollups);
```
nit: I'm somewhat nervous because we require these to be sorted and rely on the `BTreeMap` for that. The `validate_roundtrip` that we do is probably a reasonable safeguard, but maybe we want debug assertions also in `diff_field_sorted_iter()`. Could also be overkill, of course.

danhhz replied:
the debug assertion ended up super gross, but done!

```
                    let f_next = from.next();
                    debug_assert!(f_next.as_ref().map_or(true, |(fk_next, _)| fk_next > &fk));
                    f = f_next;
                    let t_next = to.next();
                    debug_assert!(t_next.as_ref().map_or(true, |(tk_next, _)| tk_next > &tk));
                    t = t_next;
```

aljoscha commented on src/persist-client/src/internal/state_diff.rs:
```diff
@@ -0,0 +1,576 @@
+                    name, single, &from
+                ));
+            }
+            assert_eq!(single, &from);
```
we're being extra certain here? 😅

danhhz replied:
lolol, artifact of development. removed

aljoscha commented on src/persist-client/src/internal/state_diff.rs:
```diff
@@ -0,0 +1,576 @@
+    match diff.val {
+        Insert(to) => {
+            let prev = map.insert(diff.key, to);
+            if prev != None {
```
All these mutate our state and then return an error if the application was incorrect. Maybe these should check before applying the diff, so as not to leave the state in an incorrect (um) state? As it is, we rely on the caller cleaning up or panic!'ing. Could also be overly cautious, though.

danhhz replied:
yeah, I thought about this, but it makes the code soooo much harder to write. given that we always panic on error, I lean toward punt. i'll drop a note here though, so future readers will know we considered this

aljoscha commented on src/persist-client/src/internal/state_versions.rs:
```diff
@@ -0,0 +1,578 @@
+
+    /// Fetches the `current` state of the requested shard.
+    ///
+    /// Uses the provided hint, which is a possibly outdated copy of all live
```
nit: `hint` comes from before the rename to `all_live_diffs`. Maybe say something like "the provided hint (`all_live_diffs`).

danhhz replied:
done

aljoscha reviewed (approved):
This is excellent! \o/ I had another round of nits, but I think we're ready to merge. (Saw that you responded to my other round of comments already and fixed some things)

danhhz commented:
TFTRs! I'm pretty burned out at this point, so just gonna merge this and then investigate the columnar format and take a stab at writing some tests on monday
